editor/utils/heartbeat.js

```diff
@@ -0,0 +1,173 @@
+ */
+import { doAction } from '@wordpress/hooks';
+
+const DISCONNECTION_NOTICE_ID = 'DISCONNECTION_NOTICE_ID';
+
+export function setupHeartbeat() {
+	const $document = jQuery( document );
+	/**
+	 * Configure heartbeat to refresh the wp-api nonce, keeping the editor authorization intact.
+	 *
+	 * @todo update the _wpnonce used by autosaves.
+	 */
+	$document.on( 'heartbeat-tick', ( event, response ) => {
+		if ( response[ 'rest-nonce' ] ) {
+			window.wpApiSettings.nonce = response[ 'rest-nonce' ];
+		}
+	} );
+
+	/**
+	 * Concatenate the title, content and excerpt.
+	 *
+	 * This is used to track changes when auto-saving.
+	 *
+	 * @since 1.9.0
+	 *
+	 * @param {Object} state The current state.
+	 *
+	 * @returns {string} A concatenated string with title, content and excerpt.
+	 */
+	const getCompareString = function( state ) {
+		return compact( [
+			getEditedPostTitle( state ),
+			getEditedPostContent( state ),
+			getEditedPostExcerpt( state ),
+		] ).join( '::' );
+	};
+
+	/**
+	 * Configure Heartbeat autosaves.
+	 */
+	const { dispatch, getState } = store;
+
+	let compareString;
+	let lastCompareString;
+
+	/**
+	 * Autosave 'save' function that pulls content from Gutenberg state. Based on `wp.autosave.save`.
+	 *
+	 * @since 1.9.0
+	 *
+	 * @returns {Object|bool} postData The autosaved post data to send, or false if no autosave is needed.
+	 */
+	const save = function() {
+		// Bail early if autosaving is suspended or saving is blocked.
+		if ( wp.autosave.isSuspended || wp.autosave._blockSave ) {
+			return false;
+		}
+
+		// Check if its time for another autosave.
+		if ( ( new Date() ).getTime() < wp.autosave.nextRun ) {
+			return false;
+		}
+
+		// Get the current editor state and compute the compare string (title::excerpt::content).
+		const state = getState();
+		compareString = getCompareString( state );
+
+		// Initialize lastCompareString if this is the first run and it is undefined.
+		if ( typeof lastCompareString === 'undefined' ) {
+			lastCompareString = compareString;
+		}
+
+		// Bail if no changes to the compare string (title::excerpt::content).
+		if ( compareString === lastCompareString ) {
+			return false;
+		}
+
+		// Store the latest compare string.
+		lastCompareString = compareString;
+
+		// Block autosaving for 10 seconds.
+		wp.autosave.server.tempBlockSave();
+
+		const postData = getCurrentPost( state );
+
+		// Dispath an event to set the state isAutosaving to true..
+		dispatch( toggleAutosave( true ) );
+
+		// Add some additional data point copies expected on the back end.
+		postData.post_id = postData.id;
+		postData.post_type = postData.type;
+
+		// Trigger some legacy events.
+		$document.trigger( 'wpcountwords', [ postData.content ] )
+			.trigger( 'before-autosave', [ postData ] );
+
+		// Trigger a hook action.
+		doAction( 'editor.beforeAutosave', postData );
+
+		// Add the nonce to validate the request.
+		postData._wpnonce = jQuery( '#_wpnonce' ).val() || '';
+
+		return postData;
```

Comment:
This is clearly an "effect", so it's kind of weird that we have to do this here but I guess The heartbeat API is already working this way in the classic editor and this avoids some extra work if we want to this the "redux" way.

Reply:
Yea, this is a bit of an artifact of glueing Gutenberg onto an existing WordPress JavaScript API

Reply:
ps. I'm attaching the nonce after triggering the events to match the existing behavior.